src/relax/transform/eliminate_common_subexpr.cc

```diff
@@ -20,223 +20,162 @@
-};
+    } else if (auto it = expr_replacements_.find(lookup_key); it != expr_replacements_.end()) {
+      VLOG(1) << "Value " << bound_value << " has previously been bound as " << it->second
+              << ".  The duplicate binding of this value to " << binding->var
+              << " will be replaced with a trivial binding, "
+              << "and occurrences of " << binding->var << " will be replaced with " << it->second;
+      output_binding = VarBinding(binding->var, it->second);
+      var_remap_.insert({binding->var->vid, it->second});
 
-class CommonSubexprEliminator : public ExprMutator {
- public:
-  explicit CommonSubexprEliminator(
-      std::unordered_map<Expr, int, StructuralHash, StructuralEqual> count_map,
-      bool call_only = false)
-      : count_map_(std::move(count_map)), call_only_(call_only) {}
+    } else {
+      VLOG(1) << "Value " << bound_value << " is bound to " << binding->var
+              << " and may be de-duplicated if it occurs again.";
 
-  // overriding here ensures we visit every subexpression
-  Expr VisitExpr(const Expr& e) override {
-    if (call_only_ && !e->IsInstance<CallNode>()) {
-      return ExprMutator::VisitExpr(e);
-    }
-    if (count_map_.count(e) && count_map_.at(e) > 1) {
-      // if we already have a mapping for it, get it
-      if (replacements_.count(e)) {
-        return replacements_.at(e);
-      }
-      // Otherwise, insert a new binding for the current expression.
-      // Visit before emitting to do inner replacements
-      Expr new_e = ExprMutator::VisitExpr(e);
-      Var v = builder_->Emit(new_e);
-      replacements_[e] = v;
-      return v;
+      expr_replacements_.insert({lookup_key, binding->var});
     }
-    return ExprMutator::VisitExpr(e);
-  }
 
-  // we are not going to do replacements inside struct info to avoid binding lots of reused shapes
-  StructInfo VisitExprDepStructInfoField(const StructInfo& struct_info) override {
-    return struct_info;
+    builder_->EmitNormalized(output_binding);
   }
 
   Expr VisitExpr_(const FunctionNode* op) override {
-    Function func = GetRef<Function>(op);
-
-    auto cache = SubexprCounter::Count(op->body);
-    std::swap(cache, count_map_);
-    Expr output = ExprMutator::VisitExpr_(op);
-    std::swap(cache, count_map_);
-
-    return output;
-  }
-
-  void VisitBinding_(const VarBindingNode* binding) override {
-    // no need to visit var def because the struct info isn't going to change
-    Expr new_value = RegisterBoundValue(binding->var, binding->value);
-
-    if (new_value.same_as(binding->value)) {
-      builder_->EmitNormalized(GetRef<VarBinding>(binding));
+    // If we have accumulated any state, visit the function in a fresh
+    // copy of the mutator, to avoid replacing a child-scope
+    // expression with a parent-scope binding, or vice versa.
+    if (expr_replacements_.size() || var_remap_.size()) {
+      return VisitWithCleanScope(GetRef<Expr>(op));
```

Comment:
I'm not sure using a clean scope in all cases is necessarily, since an inner function could capture vars from the outer scope and there might be legitimate substitutions that are possible. That said, this may not necessarily be desirable behavior since it could result in bigger closures (capturing more vars than you might want). We would also have to make sure we don't capture any DataflowVars, since that's not permitted. Any thoughts?

Reply:
Hmm.  From a lowering standpoint, my main concern would be to avoid introducing a captured variable.  If a user has an inner function without closure variables, it could be surprising if CSE prevents it from being hoisted out of the local function.

(Selfishly, I have a few plans to simplify `LambdaLift` when I have the time.  Currently, `LambdaLift` handles lifting of inner functions regardless of whether there are closure variables, which is a bit tricky to track.  I'd like to split that out into a `HoistClosureVariablesToParams` and `HoistInnerFunctions` passes.  The first would provide explicit arguments for any closure variables, and the second would hoist the inner function out to the `IRModule`.  If CSE is applied in-between those two passes, I'd like to avoid having it re-introduce closure variables.)

Reply:
Yeah I think you're right that capturing more vars might be an unwelcome surprising behavior.